### PR TITLE
fix: cache path of modules imported in setup file

### LINKED ISSUE #1450: `vi.mock` ignored if module is imported in setup file
### Describe the bug

Registration of `vi.mock` on a module will not work if that module is imported in a file added to the `setupFiles` array in `vite.config.ts`.

### Use case

We use Vue 3 with global component registration (this is needed because of the way our CMS works). To accomplish this in our tests, we register these components in `vitest.setup.ts` to the `@vue/test-utils` config as the example below.

```js
import { config } from '@vue/test-utils';

config.global.components[name] = component;
```

This causes the mocks set on the imports of these components to be ignored. Although this is our use case, this issue is unrelated to Vue. See the minimal reproduction for more details.

### Reproduction

Here is a minimal repo: https://stackblitz.com/edit/vitest-dev-vitest-etmse4?file=vitest.setup.ts

In `test/foo.test.ts` a mock is used. This fails, unless I remove the only line in `vitest.setup.ts`.

### System Info

```shell
System:
    OS: Windows 10 10.0.17763
    CPU: (4) x64 Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz
    Memory: 4.85 GB / 16.00 GB
  Binaries:
    Node: 14.17.6 - C:\Program Files\nodejs\node.EXE
    npm: 6.14.15 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Chrome: 102.0.5005.63
    Internet Explorer: 11.0.17763.2145
  npmPackages:
    @vitejs/plugin-vue: ^2.3.3 => 2.3.3
    vite: ^2.9.9 => 2.9.9
    vitest: 0.14.1 => 0.14.1
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by konpeki622**
closes #1450

### CODE REVIEW

**Comment by sheremet-va** [comment]
This doesn't seem right. `vi.mock` doesn't work because we already imported a file and it exists inside a cache. We need to remove it, if we are mocking it, that's all.

**Comment by konpeki622** [comment]
> This doesn't seem right. `vi.mock` doesn't work because we already imported a file and it exists inside a cache. We need to remove it, if we are mocking it, that's all.

I mean that once a module was imported before mocking, its path should be cached. But now it cached two different paths from setup files and test files, though these two paths refered to one module, that means it will mock 'another' module. I made them to be the same absolute path.

**Comment by sheremet-va** [comment]
> I mean that once a module was imported before mocking, its path should be cached. But now it cached two different paths from setup files and test files, though these two paths refered to one module, that means it will mock 'another' module. I made them be the same absolute path.

You are incorrect.

Their imported paths are the same. `vi.mock` mocks the correct module. It is not mocked in test, because in `request` it imports the cached one (`vite-node` doesn't know about mocks). This is why we should removed actual module from cache - it is no longer needed.

File: test/core/test/imports.test.ts [hunk 1]
```diff
@@ -23,7 +23,7 @@ test('dynamic aliased import works', async () => {
   const timeoutPath = '@/timeout'
   const variableTimeoutMod = await import(timeoutPath)
 
-  expect(stringTimeoutMod).toBe(variableTimeoutMod)
+  expect(stringTimeoutMod).toStrictEqual(variableTimeoutMod)
```
Comment (sheremet-va) [inline test/core/test/imports.test.ts:26]: No, it should `BE` the same module - have the same reference

File: packages/vitest/src/runtime/setup.ts [hunk 2]
```diff
@@ -167,7 +168,8 @@ export async function runSetupFiles(config: ResolvedConfig) {
   await Promise.all(
     files.map(async (file) => {
       getWorkerState().moduleCache.delete(file)
-      await import(file)
+      if (existsSync(file))
+        await import(file)
```
Comment (sheremet-va) [inline packages/vitest/src/runtime/setup.ts:172]: We shouldn't silently skip this

**Review by sheremet-va — CHANGES_REQUESTED** [review]

**Comment by konpeki622** [comment]
Well, I will leave this question open and find why that makes test fail. 🤔

File: packages/vite-node/src/utils.ts [hunk 3]
```diff
@@ -45,14 +45,16 @@ export function isPrimitive(v: any) {
   return v !== Object(v)
 }
 
-export function toFilePath(id: string, root: string): string {
+export function toFilePath(id: string, root: string, raw = false): string {
```
Comment (sheremet-va) [inline packages/vite-node/src/utils.ts:48]: This became very confusing. Why as a developer I should use `raw` here? This function should just work, I already gave it path

**Review by sheremet-va — CHANGES_REQUESTED** [review]
As I said, this is wrong solution. To fix the problem you are linking to, we need to load mocked version first. They have the same path. It has nothing to do with this function.

The mocked version is not loaded, because module is already cached:

https://github.com/vitest-dev/vitest/blob/b054cd5e3664055aed47a5818c7f2b79fb90ed97/packages/vite-node/src/client.ts#L85

All caches a have a key similar to `${path}__mock`.

**Pull request closed without merging**